[Shine/산토리] 로또 3단계 - 수동구매 기능 추가

zbqmgldjfh opened the pull request:
# 안녕하세요 리뷰어님! 항상 리뷰 해주셔서 감사합니다!

## 전반적인 Logic의 흐름
<img width="859" alt="스크린샷 2022-02-25 오후 4 13 33" src="https://user-images.githubusercontent.com/60593969/155671419-360a5640-8648-40a6-9114-747ca4f9ce26.png">


## 이전 feed back 반영하기
- [x] 코드 리뷰 피드백 반영하기
  - [x] `Person` 클래스가 가지고 있는 `LottoTicketSeller` 의존성 없애기
  - [x] `Person` 내부에 불필요한 멤버 변수 제거
  - [x] 6개의 연속된 번호와 보너스 번호를 따로 담는 `WinningNumbers` 객체 만들기

## 질문 1가지
저희 팀은 1급 객체로 LottoTicket을 만들었습니다.

여기서 궁금한 점은, WinningNumbers 객체를 만들면서 1급 객체인 LottoTicket을 상속하는 방법이 떠올랐습니다.
왜냐하면 기존의 1급 객체 에서의 코드를 재사용 하면 중복을 최소화 할 수 있기 때문이었습니다.

하지만 1급 객체는 독립적인 단위로 사용하는 것이다 보니, 상속을 하여 재사용 한다는 것 자체가 확실하지 않았습니다.

이에 대한 리뷰어님의 의견이 궁금합니다!

zbqmgldjfh committed:
docs(README): TodoList 작성완료

zbqmgldjfh committed:
feat(TicketFactory) TicketFactory Interface 명세 완료

zbqmgldjfh committed:
feat(RandomTicketFactory): RandomTicketFactory 생성 완료

seyoung755 committed:
feat(CustomTicketFactory): CustomTicketFactory 클래스 생성 완료

seyoung755 committed:
feat(LottoTicketSeller): Composite 방식으로 자동/수동 전환가능한 클래스 구현

- 객체 내부에 has-a TicketFactory
- Person 객체의 자동/수동 구매 메세지에 따라 구현체를 동적으로 바꿔서 동작한다.

seyoung755 committed:
feat(Person): 자동 구매/수동 구매 메서드 각각 구현

- 구매 방식에 따라 LottoTiketSeller 내부의 TicketFactory의 구현체를 변경한다.

seyoung755 committed:
modify(LottoController): 자동/수동 구매 기능 변경 중

zbqmgldjfh committed:
feat(LottoController): LottoController에서 수동 번호 받기 추가

zbqmgldjfh committed:
feat(InputView): 수동 입력 기능 추가

seyoung755 committed:
refactor(LottoController): 수동 구매 로직 분리

seyoung755 committed:
modify(InputView): 출력 함수 수정 및 close() 추가

seyoung755 committed:
docs: README TO-DO List 작성

zbqmgldjfh committed:
feat(WinningNumbers): WinningNumber Class 만들기

Lotto의 당첨번호와 보너스 넘버를 가지고 있는 Class

zbqmgldjfh committed:
modify: 함수들의 파라미터를 int[]에서 WinningNumbers로 변경 완료

seyoung755 committed:
fix: 실제 코드 변경에 따른 테스트 코드 인자 변경

seyoung755 committed:
docs: README to-do list 결과 반영

seyoung755 committed:
feat: 사용자 입력값에 대한 검증 로직 추가

  - [x] 당첨번호에 대한 검증
    - [x] 당첨번호의 개수
    - [x] 당첨번호의 범위
    - [x] 보너스 번호의 검증
  - [x] 수동 로또 개수에 대한 검증
  - [x] 당첨 번호 간 중복 처리
  - [x] 보너스 번호와 당첨 번호 간 중복 처리

seyoung755 committed:
modify: 불필요한 메서드 제거

seyoung755 committed:
style: formatting 수정

seyoung755 committed:
docs: README to-do list 수정

seyoung755 committed:
modify: Person 멤버 변수 중 필요 없는 name 필드 제거

seyoung755 committed:
modify: Person 내부 LottoTicketSeller 의존성 제거

- Person 내부에서 LottoTicketSeller를 통해 티켓을 구매하는 기능을 제거하였습니다.
- 해당 기능을 LottoController에서 처리하도록 하였습니다.
- Person의 money 필드도 제거하였습니다.

zbqmgldjfh committed:
test(LottoTicket): LottoTicket 수동 검증 완료

zbqmgldjfh committed:
refactor(LottoController): LottoController 리팩토링 완료

zbqmgldjfh committed:
refactor(LottoController): LottoController 리팩토링 완료

zbqmgldjfh committed:
modify(InputView.requestMoney): 사용자의 입력 금액 검증 로직 추가

zbqmgldjfh committed:
test(PersonTest): Person의 당첨 번호 검증로직 수정

zbqmgldjfh committed:
docs(README): 피드백 반영 사항 확인 완료

seyoung755 committed:
modify(LottoTicketSeller): 불필요한 Setter 제거

seyoung755 committed:
refactor(LottoTicket): LottoTicket 내부 인덴트 줄이기 완료

- 인덴트 깊이가 2인 메서드를 1로 수정하였습니다.

seyoung755 committed:
modify(InputView): 번호 입력 시 구분자 통일

- 수동 구매 로또 번호와 당첨번호 입력 시 구분자를 통일하였습니다.

wheejuni commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,48 +1,82 @@
+    private void buyRandomLottoTicket(Person customer, int money) {
+        LottoTicketSeller lottoTicketSeller = new LottoTicketSeller(new RandomTicketFactory());
+
+        int ticketCount = money / TICKET_PRICE;
```
`seller` 가 금액을 보고 알아서 몇 장 내줄지 판단하게 하는 건 어떨까요?

zbqmgldjfh replied:
원래 Step2 에서는 Seller 가 스스로 몇장을 구매할지 판단하게 되어 있었습니다.
기존 Person은 Seller에 대한 의존성이 있어 Person이 Seller에게 돈을 넘겨주고, 이를 Seller가 스스로 인지하여 티켓을 판매하습니다.
하지면 변경 도중 바뀌게 되었는데, 그 이유는 2가지 입니다.

----
**전략 패턴 사용중 시그니처의 통일**

Composition을 활용한 전략패턴을 Seller가 사용하고 있는데, 이때문에 exchangeTicket 의 시그니처를 random 과 custom 구매 방식에서 동일하게 가게 되었습니다.

수동 구매하기는 사용자의 번호를 전달하면 하나의 LottoTicket을 반환하는 Query 메서드 입니다.
자동 구매는 스스로 Random한 번호를 알고있는 Query 메서드 입니다.
즉, 둘 간의 메서드 시그니처를 동일하게 하다보니, 인자로 Money를 전달할수가 없었습니다.

Random쪽에 파라미터를 만들어 돈을 넘겨주면, 수동 구매 쪽에도 불필요한 파라미터를 만들어야 하기 때문입니다.
문제는 수동은 매번 사용자의 번호를 받아 LottoTicket을 만들어야 하기 때문입니다.

----
**Seller를 생성할 당시에 전달해야 하는가?**

그럼 2번째 방법으로 Seller를 생성할 시점에 인자로 돈을 전달하는 방법이 있을탠데...
Seller가 사용자가 가지고 있는 돈에 대한 정보를 꼭 알아야 할까? 란 의문이 있었습니다.

그냥 seller는 요청할때마다 반환해주면 되는것 아닐까? 곡 Seller도 자신의 상태값으로 money를 가지고 있어야 하나? 란 의문이 있었던것 같습니다.

wheejuni commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,48 +1,82 @@
-        return temp;
+    private ArrayList<LottoNumber> getLottoNumbers() {
+        int[] numbers = InputView.requestCustomTicketNumber();
+        ArrayList<LottoNumber> lottoNumbers = new ArrayList<>();
```
꼭 `ArrayList` 타입으로 명시해줄 필요가 있었는지 궁금합니다!

seyoung755 replied:
저희도 프로젝트 진행하면서 해당 부분이 계속 마음에 걸렸습니다..! 제가 미션의 요구사항 중에 ArrayList를 사용해 구현하라는 것을 조금 잘못 이해해서(지금 생각해보니 List의 구현체를 ArrayList로 쓰라는게 아니라 배열 대신 사용해보라는 뜻이었던 것 같습니다.) ArrayList 형태로 LottoNumber를 저장하는 과정에서 단추가 잘못끼워져서 많은 곳에 혼재되었던 것 같습니다. 

리턴 타입이나 매개변수 타입은 List로 선언하여 구현체가 바뀌더라도 설계에 유연성이 생기도록 하는 것이 바람직하다는 걸 참고링크를 보고 다시금 배운 것 같습니다!

zbqmgldjfh replied:
음 요구사항 자체에 "ArrayList를 사용하라" 라는 사항이 있어 이를 준수하기 위해 사용한 것 같습니다.
원래대로 였다면 List를 사용했을 것 같습니다. 

팀원 말대로 단순 1차원 배열 대신 ArrayList를 사용하라는 의미가 아니였을까? 란 생각을 다시 하게 되는것 같습니다.

wheejuni replied:
요구사항에  _`ArrayList` 를 사용하라_ 고 되어있다면 앞으로는 `java.util.List` 를 사용하라 라는 의미로 해석해주시면 좋을 것 같습니다.

wheejuni commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,48 +1,82 @@
+        return lottoNumbers;
+    }
+
+    private int[] makeResult(Person customer) {
```
꼭 배열을 사용할 이유가 있었는지 궁금합니다!

seyoung755 replied:
맞은 개수 별로 result를 저장하기 위해서였습니다. 단순히 맞은 개수의 경우의 수만큼 배열을 선언해두면 index에 접근하기 편해서 배열을 사용했었습니다. 지금 생각해보니 그런 방식으로 구현하더라도 HashMap 등의 선택지도 있었다는 생각이 드네요!

zbqmgldjfh replied:
저희 팀이 배열을 사용한 이유는! 각 index 별로 알맞은 값을 추가해주면, 예를 들어 

0번 index 에 5개 => 0개 맞은 사람이 5명
1번 index 에 2개 => 1개 맞은 사람이 2명
...
6번 index에 1개 => 6개 모두 맞은 사람이 1명
이런 식으로 간단하게 표현할 수 있다 생각하였기 때문입니다.

혹, 리뷰어님 께서 다른 자료구조를 생각한 것 이라면? 어떤 이유가 있는 것 일까요?
배열로 의미를 잘 전달할 수 있다 생각했는데, 더 나아가 따로 Result라는 Class로 만들어야 했을까요?

wheejuni commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -42,7 +42,7 @@ public int compareTo(LottoNumber o) {
 
-    public boolean isSameNumber(int number) {
-        return this.number == number;
+    public boolean isSameNumber(LottoNumber lottoNumber) {
+        return this.number == lottoNumber.number;
```
`.equals()` 를 오버라이드하는 판단이 더 좋았을 것 같습니다.

seyoung755 replied:
넵,, 심지어 이미 equals()를 사용하고 있기도 했는데 놓쳤네요. 피드백 감사합니다!

zbqmgldjfh replied:
음... 원래 equals를 인식은 했는데...(바로 위에 있으니 ㅠ,ㅠ) 좀더 메시지에 의미부여를 하고싶어 따로 만들었습니다.
1) number.equals(another)
2) number.isSameNumber(another)
음... 돌이켜 생각해보니 equals로도 의미가 충분하게 전달되는 것 같습니다. ㅎㅎ!!

wheejuni commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -24,19 +25,25 @@ private boolean isValidTicketSize(ArrayList<LottoNumber> lottoNumbers) {
-            }
+    private int checkWinningNumber(WinningNumbers winningNumbers, LottoNumber myNumber) {
+        int result = 0;
+        for (int i = 0; i < winningNumbers.getSize(); i++) {
+            result += check(winningNumbers.getNumberOfIndex(i), myNumber);
+        }
+        return result;
```
스트림 API로 개선할 수 있을 것 같네요.
`.filter()`, `.count()` 를 사용하면 될 것 같아요. 필요에 맞는 종단 처리 메소드가 거의 다 제공되므로 다양한 시나리오에 맞게 잘 찾아서 사용해 보시면 좋겠어요.

seyoung755 replied:
이 부분은 인덴트를 줄이기 위해 나중에 메서드를 분리해서 작성했었는데 그런 방법이 있다는 것도 알게되었네요. 말씀해주신대로 `Intstream`을 활용한다고 생각하니 더 간결해보이는 것 같습니다!

wheejuni commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -24,19 +25,25 @@ private boolean isValidTicketSize(ArrayList<LottoNumber> lottoNumbers) {
+        return result;
+    }
+
+    private int check(LottoNumber numberOfIndex, LottoNumber myNumber) {
```
위에서 알려드린 방법대로 리팩터링하려면, 여기선 `boolean` 을 리턴해야겠습니다.

zbqmgldjfh replied:
아하! 말씀해주신 대로 boolean을 사용하는 것 이 check 라는 메서드의 의미에도 더 적합하다 생각되는 것 같습니다.

wheejuni reviewed (changes requested):
수고 많으셨고 전반적으로 잘 구현해 주셨습니다~ 💯 
코멘트 참고해 주시고요.
플젝 전반에서 `List` 와 `ArrayList` 타입 선언이 혼재되어 굉장히 어지러운 모습을 보여주고 있는데요,
[이 링크](https://stackoverflow.com/questions/9852831/polymorphism-why-use-list-list-new-arraylist-instead-of-arraylist-list-n) 한번만 참고해주세요~

wheejuni commented:
질문은 별도 코멘트로 답변 드릴게요~

wheejuni commented:
중복을 최소화하는 방법으로의 상속 사용에 별로 **동의하지 않습니다**
상속 자체가 안티패턴이라고 보는 시각도 있고요, 코틀린에는 `data class` 가 있는데 `data class` 는 `open class` 가 될 수 없어서 상속받을 수조차 없습니다.

만약 중복을 걷어내고자 하는 방법이라면 상속보다는 합성을 고려해 보시는게 좋겠고요,
비슷한 고민을 담은 [이 글](https://hue-dev.site/engineering/2021/05/20/%EC%83%81%EC%86%8D%EA%B3%BC-%ED%95%A9%EC%84%B1%EC%97%90-%EA%B4%80%ED%95%98%EC%97%AC.html) 을 읽어보시고 해결방안을 고민해 보시면 좋겠어요~ 감사합니다.

zbqmgldjfh committed:
docs(README): Review 변경사항 to-do list 작성완료

zbqmgldjfh committed:
refactor(LottoNumber): 기존의 isSameNumber 메서드를 equals 로 대체하기

- 변경 전
number.isSameNumber(another) 과 같이 사용

- 변경 후
number.equals(another) 과 같이 사용하도록 equals로 수정

zbqmgldjfh committed:
refactor(LottoTicket): 기존의 for문들을 stream으로 변경하기

zbqmgldjfh committed:
docs(README): 요구사항 수정 완료

seyoung755 committed:
refactor(LottoTicket.check): if문을 제거하고 바로 검사 결과를return하도록 수정

seyoung755 committed:
refactor(LottoTicket): 불필요하게 ArrayList로 명시된 부분을 List로 수정

seyoung755 committed:
remove(LottoTicketFactory): 기존에 사용하던 팩토리 클래스 삭제

seyoung755 committed:
test(TicketFactory): 팩토리 클래스 변경에 따른 테스트 수정

seyoung755 committed:
refactor: 불필요한 ArrayList 명시를 List로 수정 완료

- LottoTicketSeller와 TicketFactory 간 티켓을 주고받을 때,
불필요하게 ArrayList로 명시하여 유연성을 떨어뜨리는 설계를 수정하였습니다.

zbqmgldjfh committed:
docs: 그림 추가

그림 추가
